Add support for user-defined healthchecks

talex5 opened the pull request:
This PR adds support for user-defined health-check probes for Docker containers. It adds a  `HEALTHCHECK` instruction to the Dockerfile syntax plus some corresponding "docker run" options. It can be used with a restart policy to automatically restart a container if the check fails.

The `HEALTHCHECK` instruction has two forms (more may be added later, e.g. `HTTP`):
- `HEALTHCHECK [OPTIONS] CMD command` (check container health by running a command inside the container)
- `HEALTHCHECK NONE` (disable any healthcheck inherited from the base image)

The `HEALTHCHECK` instruction tells Docker how to test a container to check that it is still working. This can detect cases such as a web server that is stuck in an infinite loop and unable to handle new connections, even though the server process is still running.

When a container has a healthcheck specified, it has a _health status_ in addition to its normal status. This status is initially `starting`. When a health check passes, it becomes `healthy`. After a certain number of failures, it becomes `unhealthy`.

The options that can appear before `CMD` are:
- `--interval=DURATION` (default: `30s`)
- `--timeout=DURATION` (default: `30s`)
- `--grace=DURATION` (default: `30s`)
- `--retries=N` (default: `1`)
- `--exit-on-unhealthy=X` (default: `true`)

The health check will first run **interval** seconds after the container is started, and then again **interval** seconds after each previous check completes.

If a single run of the check takes longer than **timeout** seconds then the check is considered to have failed.

If the health state is `starting` and a check started within **grace** seconds of the container's start time fails, the failure is ignored and the health state remains `starting`.

It takes **retries** consecutive failures of the health check for the container to be considered `unhealthy`.

If `--exit-on-unhealthy` is `true` then the container will exit as soon as it becomes unhealthy. The container may then be automatically restarted, depending on its restart policy.

For example, to check every five minutes or so that a web-server is able to serve the site's main page within three seconds:

```
HEALTHCHECK --interval=5m --grace=20s --timeout=3s --exit-on-unhealthy \
  CMD curl -f http://localhost/
```

(from https://github.com/talex5/docker/blob/healthcheck/docs/reference/builder.md#healthcheck)

The changes to "docker run" are described here:

https://github.com/talex5/docker/blob/healthcheck/docs/reference/run.md#healthcheck

**Example**

```
$ docker run --name=test -d \
    --health-cmd='stat /etc/passwd' \
    --health-interval=2s \
    --exit-on-unhealthy=false \
    busybox sleep 1d
$ sleep 2; docker inspect --format='{{.State.Health.Status}}' test
healthy
$ docker exec test rm /etc/passwd
$ sleep 2; docker inspect --format='{{json .State.Health}}' test
{
  "Status":"unhealthy",
  "FailingStreak":1,
  "LastCheckStart":"2016-05-09T11:09:09.673709108Z",
  "LastCheckEnd":"2016-05-09T11:09:09.786146142Z",
  "LastExitCode":1,
  "LastOutput":"stat: can't stat '/etc/passwd': No such file or directory\n"
}
```

The health status is also displayed in the `docker ps` output.

**Description for the changelog**: Add support for user-defined healthchecks

Closes #21142 and #21143.

Commit:
Add HEALTHCHECK feature

Signed-off-by: Thomas Leonard <thomas.leonard@docker.com>

thaJeztah commented:
thanks @talex5!

ping @icecrime :smile:

cpuguy83 commented on docs/reference/builder.md:
```diff
@@ -1291,6 +1291,68 @@ The `STOPSIGNAL` instruction sets the system call signal that will be sent to th
+
+The options that can appear before `CMD` are:
+
+* `--interval=DURATION` (default: `30s`)
```
These all seem to be runtime configurations that ought not be in the Dockerfile.

justincormack replied:
It does seem a bit problematic. I might test my container on a fast unloaded machine with SSD and then you run it on a slow loaded machine with a hard drive and the startup time is much longer so it fails the health check (eg mysql seems to take a long time to start serving I noticed recently). It does seem odd to have values that are related to runtime performance in the dockerfile.

stevvooe replied:
`delay`?

cpuguy83 replied:
I think I'm really a hard -1 on these flags in the Dockerfile.

cpuguy83 commented on docs/reference/builder.md:
```diff
@@ -1291,6 +1291,68 @@ The `STOPSIGNAL` instruction sets the system call signal that will be sent to th
+
+The `HEALTHCHECK` instruction has two forms:
+
+* `HEALTHCHECK [OPTIONS] CMD command` (check container health by running a command inside the container)
```
Having `CMD` here is really confusing.

I think it would be better have simply `HEALTHCHECK <cmd>` that would be the default health checking command, kind of like `ENTRYPOINT` but for healthchecks.

talex5 replied:
There will likely be other types in future. e.g. `HEALTHCHECK HTTP port=80` or somesuch. Without the `CMD`, it will be hard to extend the syntax.

tianon replied:
Maybe `RUN` instead?  Or go with simpler lowercase "types" (like "run") so that they clash less visually with other instructions?

cpuguy83 replied:
Ok, I understand the case for `CMD` here.

cpuguy83 commented on vendor/src/github.com/docker/engine-api/types/container/config.go:
```diff
@@ -5,6 +5,20 @@ import (
 )
 
+// HealthConfig holds configuration settings for the HEALTHCHECK feature.
+// For durations, a value of -1 means to inherit the default value.
```
Instead of the magic number, can the durations be pointers? `nil` Means inherit.

talex5 replied:
Done. Optional floats are now represented by `nil`.

talex5 replied:
At the request of @stevvooe I've removed the pointers again. Now that `grace` isn't an option, 0 is no longer a valid duration for any option and we can use zero for inherit.

cpuguy83 commented on vendor/src/github.com/docker/engine-api/types/container/config.go:
```diff
@@ -24,6 +38,7 @@ type Config struct {
 	StdinOnce       bool                  // If true, close stdin after the 1 attached client disconnects.
 	Env             []string              // List of environment variable to set in the container
 	Cmd             strslice.StrSlice     // Command to run when starting the container
+	Healthcheck     *HealthConfig         `json:",omitempty"` // Command to run to check container is healthy
```
Only 1 health check per container? Do you see a reason why someone would want more than 1?

I can think at least testing multiple listening ports.

thaJeztah replied:
Wouldn't those be combined? (just thinking)

justincormack replied:
You could with a script, but I guess just using curl as your healthcheck is quite nice.

crosbymichael replied:
Why does this have an `omitempty`?  How does it get marshaled to json and sent to the server?

stevvooe replied:
@justincormack I would expect support for multiple `HEALTHCHECK` clauses in a Dockerfile. Any reason not to?

justincormack replied:
On 13 May 2016 21:20, "Stephen Day" <notifications@github.com> wrote:
>
> In vendor/src/github.com/docker/engine-api/types/container/config.go:
>
> > @@ -24,6 +38,7 @@ type Config struct {
> >  StdinOnce       bool                  // If true, close stdin after
the 1 attached client disconnects.
> >  Env             []string              // List of environment variable
to set in the container
> >  Cmd             strslice.StrSlice     // Command to run when starting
the container
> > + Healthcheck     *HealthConfig         `json:",omitempty"` // Command
to run to check container is healthy
>
> @justincormack I would expect support for multiple HEALTHCHECK clauses in
a Dockerfile. Any reason not to?
>

How would you then eg override one of them? Maybe it should be an array of
checks so you have to replace all of them?

stevvooe replied:
I see complexity this can bring on. I think the issue is that you may want to define a tcp, http and cmd based healthcheck. Perhaps, it is sufficient to let these coexist in a single command, eliding the need to have complex merge semantics of health declarations (and having to name the health checks to tell them apart).

talex5 replied:
@crosbymichael `omitempty` only omits the value if it's empty (nil). If it contains information, it will be converted to JSON.

crosbymichael replied:
@talex5 thanks, i blanked out on this one, its correct

cpuguy83 commented:
There doesn't seem to be anything documented on the protocol between docker and the probe.
I imagine this is just exit statuses, but would be good to document.

Do we want this to be a simple 0 == OK, not zero == bad or more robust ala nagios-style checks (ok, warning, critical, and status messages for reporting to the user)?

cpuguy83 commented:
Images also often define the ports they listening on (`EXPOSE`), would we want to provide a flag that runs some default check on these ports (like open a TCP conn)?  This would be similar to how `-P` automatically creates port forwards for all `EXPOSE`'d ports.

crosbymichael commented on daemon/health.go:
```diff
@@ -0,0 +1,310 @@
+
+// exec the healthcheck command in the container.
+// Returns the exit code and error message (if any)
+func (*cmdProbe) run(ctx context.Context, d *Daemon, container *container.Container) (int, string) {
```
In go, your second return value should be `error` here and not return strings

talex5 replied:
The error here is normally the error string returned by the probe rather than a Go error, although there are a few places where I handle the (unlikely) case of a Go error starting the probe by pretending that the error came from the probe itself. If we return these errors as a third return value, we'll just turn it into a string at the next opportunity anyway.

stevvooe replied:
@talex5 Errors in Go are just values. You can add the error content to the error that makes sense for the application. I would expect these to build on `exec.ExitError` or define a special error type to encapsulate the error. Use a stringly-typed approach is just calling for bugs.

crosbymichael replied:
Ya, this return value should be an `error` 100%.  Even if it is just a string you can use `errors.New("my string")` or `fmt.Errorf("my %s", "string")` but you should not return `string` especially for how you are using it.

crosbymichael commented on vendor/src/github.com/docker/engine-api/types/container/config.go:
```diff
@@ -5,6 +5,20 @@ import (
+	Timeout     float64           // Time to wait before considering the check to have hung.
+	Retries     uint              // Number of consecutive failures needed to consider a container as unhealthy.
+				      // 0 means default.
+	ExitOnUnhealthy *bool	      `json:",omitempty"` // Stop container on entering unhealthy state. nil for default.
```
Is there a reason why this is a nilable option?  They way I see it is that its a yes or no entry and it being nil does not mean anything.  I think it would be better if this could not be nil.

stevvooe replied:
I'm not sure this option should even be present...

Either way, default of `false` is sensible, so no need for the pointer.

talex5 replied:
(note: @aluzzardi asked for the default to be changed to `true`)

An image can turn this on or off. The user can set it on or off or inherit the default from the image, so we need three states here.

crosbymichael commented on container/health.go:
```diff
@@ -0,0 +1,63 @@
+// OpenMonitorChannel creates and returns a new monitor channel. If there already is one,
+// it returns nil.
+func (s *Health) OpenMonitorChannel() chan struct{} {
+	if s.stop == nil {
```
This is racy, you will need to protect this with a lock

talex5 replied:
Opening and closing the channel only happens from the event monitor thread, which already runs with the container locked.

crosbymichael replied:
ok

crosbymichael commented on container/health.go:
```diff
@@ -0,0 +1,63 @@
+		// This channel does not buffer. Once the write succeeds, the monitor
+		// has read the stop request and will not make any further updates
+		// to c.State.Health.
+		s.stop <- struct{}{}
```
You can just do a `close(s.stop)` then set it to nil instead of sending an empty struct

talex5 replied:
Using `close` will not ensure that the probe thread has stopped before continuing, leading to a possible race if it is just about to write to the health structure.

crosbymichael replied:
Ok, i see how its being used

stevvooe commented on api/server/router/container/backend.go:
```diff
@@ -33,14 +33,14 @@ type copyBackend interface {
 type stateBackend interface {
 	ContainerCreate(types.ContainerCreateConfig) (types.ContainerCreateResponse, error)
-	ContainerKill(name string, sig uint64) error
+	ContainerKill(name string, sig uint64, cancelRestartManager bool) error
```
Make an options structure for this.

talex5 replied:
Done.

stevvooe commented on api/server/router/container/backend.go:
```diff
@@ -33,14 +33,14 @@ type copyBackend interface {
 	ContainerRm(name string, config *types.ContainerRmConfig) error
 	ContainerStart(name string, hostConfig *container.HostConfig) error
-	ContainerStop(name string, seconds int) error
+	ContainerStop(name string, seconds int, cancelRestartManager bool) error
```
Make an options structure for this.

stevvooe replied:
Or at least document what `cancelRestartManager` means.

stevvooe commented on builder/dockerfile/builder.go:
```diff
@@ -22,15 +22,16 @@ import (
 )
 
 var validCommitCommands = map[string]bool{
```
Lose the bool type here and save some space: use `struct{}`.

talex5 replied:
The PR does not change this code. If you want to clean up nearby code, I it think should be done in a separate PR.

stevvooe commented on vendor/src/github.com/docker/engine-api/types/container/config.go:
```diff
@@ -5,6 +5,20 @@ import (
+// HealthConfig holds configuration settings for the HEALTHCHECK feature.
+// For durations, a value of -1 means to inherit the default value.
+type HealthConfig struct {
+	Test        strslice.StrSlice // Test to perform to check container is healthy.
```
`Kind`, `Type`, `Mode`. `Test` doesn't really work here.

talex5 replied:
It's not just the kind. It's something like `{"CMD", "check.sh"}`,

stevvooe replied:
That is no good. We don't want to multi-band this slice. There should be a separate type and data field.

stevvooe commented on vendor/src/github.com/docker/engine-api/types/container/config.go:
```diff
@@ -5,6 +5,20 @@ import (
+	GracePeriod float64           // Time after container start to assume failure means still starting.
+	Interval    float64           // Time to wait between checks.
+	Timeout     float64           // Time to wait before considering the check to have hung.
+	Retries     uint              // Number of consecutive failures needed to consider a container as unhealthy.
```
No need for unsigned, just use `int`.

talex5 replied:
Why? Then I'd have to handle negative values somehow, and it's not clear what that would mean.

stevvooe replied:
You don't. There is no need for `uint` here. It just makes other code awkward.

stevvooe commented on vendor/src/github.com/docker/engine-api/types/container/config.go:
```diff
@@ -5,6 +5,20 @@ import (
+	Test        strslice.StrSlice // Test to perform to check container is healthy.
+				      // An empty slice means to inherit the default.
+				      // Otherwise, Cmd[0] is "NONE" or "CMD".
+	GracePeriod float64           // Time after container start to assume failure means still starting.
```
Please use `time.Duration`.

talex5 replied:
Unfortunately, Go converts durations to JSON by printing its internal format (integer nanoseconds), which would leak an implementation detail of the Go language into the protocol, so I converted to floating-point seconds instead as a more standard unit.

stevvooe replied:
Please use `time.Duration`. The integer nanoseconds format is portable and the `time.Duration` type's behavior is safe and well-known.

stevvooe commented on vendor/src/github.com/docker/engine-api/types/types.go:
```diff
@@ -274,6 +274,16 @@ type ExecStartCheck struct {
+type Health struct {
+	Status         string    // States are: "starting", "healthy", "unhealthy"
+	FailingStreak  uint64    // Number of consecutive failures
+	LastCheckStart time.Time // Zero if never started
```
Create subtype `Last`, with these fields:

```go
type HealthcheckResult {
  Start, End time.Time
  ExitCode int
  Output []byte // string may not be right type.
}
```

Then, use a slice for the latest n results:

```
type Health struct {
  Results []HealthCheckResult
}
```

Contents of results can be managed by policy or tweaked for ideal use case. Things like last N failures can be calculated by looking at entries.

stevvooe replied:
Also, may want to include an `Err error` field to keep track of any errors resolving health.

stevvooe commented on docs/reference/builder.md:
```diff
@@ -1291,6 +1291,74 @@ The `STOPSIGNAL` instruction sets the system call signal that will be sent to th
+* `--timeout=DURATION` (default: `30s`)
+* `--grace=DURATION` (default: `30s`)
+* `--retries=N` (default: `1`)
+* `--exit-on-unhealthy=X` (default: `true`)
```
We can probably lose the preposition here. In general, I find the wording on this field to be awkward but I don't have a better suggestion.

stevvooe replied:
After some thought, this policy really belongs on the restart policy. The health check should just be about reporting status. The interpretation of the status should be part of the restart policy.

stevvooe commented on builder/dockerfile/parser/testfiles/health/result:
```diff
@@ -0,0 +1,9 @@
+(from "debian")
```
Disappointing that s-expressions have worked their way in. These should be table based tests but that seems to be out of scope of this review.

mgoelzer commented:
ping @tianon

tianon commented:
I'm a little worried that we're pigeonholing ourselves by specifying what to do with "unhealthy" containers so early in the life of this feature, especially allowing that to be baked into the image.  Some folks are going to want more complex logic for that, or simply a dashboard of status, and IMO the default should be to do nothing extra with this information besides collect it.

talex5 commented:
@cpuguy83 Good idea - I've added documentation for the probe API now.

talex5 commented on runconfig/opts/parse.go:
```diff
@@ -351,6 +389,62 @@ func Parse(cmd *flag.FlagSet, args []string) (*container.Config, *container.Host
+			if runtime.GOOS != "windows" {
+				args = []string{"CMD", "/bin/sh", "-c", *flHealthCmd}
+			} else {
+				args = []string{"CMD", "cmd", "/S", "/C", *flHealthCmd}
```
This isn't right. Which shell to use should be decided by the daemon, not the client. But I'm not sure what to do here. The other commands in `docker run` (`--entrypoint` and the `run` command) don't use the shell at all. The easiest thing would be to make it work like entrypoint, but that might be inconvenient.

The other option is the change the format to allow e.g. `{"CMD-SHELL", "check.sh", "-q"}` to be sent over the wire, and have the daemon convert to the appropriate shell. Any preference?

stevvooe commented on vendor/src/github.com/docker/engine-api/types/container/config.go:
```diff
@@ -5,6 +5,27 @@ import (
+	Test strslice.StrSlice `json:",omitempty"`
+
+	// Nil means to inherit.
+	GracePeriod *float64 `json:",omitempty"` // Time after container start to assume failure means still starting.
```
Let's make sure this feedback doesn't get lost:

1. These _must_ be `time.Duration`. This is how time types are represented in Go.
2. Use the zero-value to communicate meaning. For example, in case these are the zero-value, use a default value. There are no need for pointers here.

If the issue is serialization, create a `time.Duration`-compatible that exports methods for conversion to `time.Duration`. However, I see no issue in serialization these an nanosecond-integers.

talex5 replied:
If you do `--health-grace=0` then you want a grace period of zero, which is not the same as not specifying the grace period (which inherits from the image).

stevvooe replied:
I'd really recommend avoiding trying to define an inheritance scheme for doing overrides. Just place these into a separate structure to allow overriding the parameters for the check or the just the test. Placing pointers across these fields is unnecessary. Look around in this file; they aren't there.

Looking back, I am trying to find the PR against engine-api for these changes. It looks like this needs to go for further review.

stevvooe commented on vendor/src/github.com/docker/engine-api/types/container/config.go:
```diff
@@ -5,6 +5,27 @@ import (
+
+	// Stop container on entering unhealthy state.
+	// Nil means inherit.
+	ExitOnUnhealthy *bool `json:",omitempty"`
```
Again, no need for pointer here.

talex5 replied:
If you lose the pointer then you have to throw away one of the current options:

- Use policy specified in the image (`docker run image`).
- Exit on unhealthy, whatever the image says (`docker run --exit-on-unhealthy image`).
- Do not exit on unhealthy, whatever the image says (`docker run --exit-on-unhealthy=false image`).

Which are you proposing to drop?

stevvooe replied:
Using a boolean pointer in Go is an anti-pattern.

There are only two options. Exit on unhealthy or not. The default is to not (bools are initilized to false by default). Use ordering to achieve this behavior:

1. Grab image health check. If image specifies to exit, set to true.
2. Apply flags. Set based on flag.

None of this requires a pointer. All you need to know is whether or not the flag was set. If it was, apply it. Adding a pointer here just makes a mess.

Either way, this belongs on restart policy, rather than the health check.

crosbymichael commented on docs/reference/builder.md:
```diff
@@ -1291,6 +1291,74 @@ The `STOPSIGNAL` instruction sets the system call signal that will be sent to th
+* `--interval=DURATION` (default: `30s`)
+* `--timeout=DURATION` (default: `30s`)
+* `--grace=DURATION` (default: `30s`)
+* `--retries=N` (default: `1`)
```
Because of this retries, does this mean that a container can never go from unhealthy to healthy again?  What happens if there is a network blip and a single check fails but then the network comes back later?  Is my container stuck in unhealthy for the rest of its life until we restart it or its auto killed?  

The whole retry count and exit on unhealthy seems very... intense for a production level feature.  Fail one check and the container dies...

@talex5 Mostly commenting for ppl that helped designed this to weigh in.  ;)

talex5 replied:
I've clarified the text. It now reads:

> When a container has a healthcheck specified, it has a _health status_ in
 addition to its normal status. This status is initially `starting`. When a
health check passes, it becomes `healthy` (whatever state it was previously in).
After a certain number of consecutive failures, it becomes `unhealthy`.

HackToday replied:
OK. seems it could not change from unhealthy to healthy status ? Your above comment said only from starting--> heatlthy, or starting-->unhealthy.
What about other status transition, like unheatlty ---> healthy or vice verse.

dongluochen commented on builder/dockerfile/parser/testfiles/health/Dockerfile:
```diff
@@ -0,0 +1,10 @@
+ADD check.sh main.sh /app/
+CMD /app/main.sh
+HEALTHCHECK
+HEALTHCHECK --interval=5s --timeout=3s --retries=1 --grace=4s --exit-on-unhealthy \
```
Generally speaking retries should be multiple like 3 to avoid system instability. We don't want to provide examples of retries=1.

talex5 replied:
This is a unit-test for the parser. It's not intended as an example for users (it also specifies mutliple healthchecks and other unreasonable things).

dongluochen commented on daemon/health.go:
```diff
@@ -0,0 +1,312 @@
+	defaultGracePeriod = 30 * time.Second
+
+	// Shut down a container if it becomes Unhealthy.
+	defaultExitOnUnhealthy = true
```
I prefer `defaultExitOnUnhealthy` to be no action.

talex5 replied:
@aluzzardi Could you comment here? I think you asked for the default to be `true`.

stevvooe replied:
Also, I really think this belongs on the restart policy.

dongluochen commented on docs/reference/builder.md:
```diff
@@ -1291,6 +1291,74 @@ The `STOPSIGNAL` instruction sets the system call signal that will be sent to th
+The `HEALTHCHECK` instruction has two forms:
+
+* `HEALTHCHECK [OPTIONS] CMD command` (check container health by running a command inside the container)
+* `HEALTHCHECK NONE` (disable any healthcheck inherited from the base image)
```
I don't think there should be inheritance in health check. Implicitly running a health check from a base image seems wrong.

cpuguy83 replied:
I'm inclined to agree with this. It seems powerful, but also annoying.
And nulling out a healthcheck should really be addressed along with other declared items as well (ENV, VOLUME, EXPOSE, etc.)

mgoelzer replied:
I'm coming around to agree with you.  Inheritance may be confusing to image creators who might not expect this behavior.

@crosbymichael @icecrime

thaJeztah replied:
Otoh, I can see "official" images having a proper health check set, not inheriting those  would put extra load on people consuming those images (I agree with needing a way to unset things though) /cc @tianon

tianon replied:
Indeed, this is one of many reasons that I don't think the health checks should do much more than collect status by default unless the operator who is starting the container explicitly requests that they have some effect on the container's lifecycle -- if that were the case, then inheriting a health check should be mostly harmless (although a way to "reset" and remove the health check would obviously be necessary).

crosbymichael replied:
Like @tianon said, maybe the first iteration/merge of this feature should probably only have health checking as a status for many reasons.  First we want to make sure it is correct and robust as this is a production feature.  

Second, having actions in docker that are taken on the health check results promotes people to open PRs to docker to add all the functionality that they want.  If heath checks were just a status and we provide a way to look at the last X results, it promotes people, etc to build their logic for acting upon the results outside of Docker.  

Its not that we wouldn't want to code actions into docker, maybe restart policies are the best place to add the exit on failure, but there is no way we can support all of needs of people who want to do something on a health check result.  Having a robust way to see the status, past and present is a better long term solution for this.  

Also the current defaults for retry and exit are very extreme.  A container cannot go from healthy to unhealthy and back to health right now without being killed.  Just because you have a network blip once does not mean that your database should be killed and restarted.  

Changing the defaults to something less extreme would help if we want to support inheritance.

jfrazelle replied:
Agreed w @tianon

stevvooe commented:
Were the datastructure changes ever reviewed in engine-api? I don't see an edit to the vendor scripts and cannot find a PR with these changes. Considering the current issues with the data structures, I think we need to have a step back.

cpuguy83 commented:
@steevvoe you are looking at the discussion. Design discussion happening here, if approved a pr is typically  opened against engine-api and vendor gets updated here (assuming engine-api is approved)

stevvooe commented:
@cpuguy83 SGTM

cpuguy83 commented:
Looks like this was moved to code-review when really the design is not ready, so I'm moving back to design-review.

cpuguy83 commented:
To sum up the discussion so far (please correct if this is not accurate):

1. It seems like we don't want to support inheritance out of the gate... so it sounds like this should not be in the Dockerfile at all
2. We don't want to handle failure cases within Docker right now, only collect and expose the information in the API

talex5 commented:
@cpuguy83 This makes sense to me. If noone involved in the original design objects, I'll remove the `exit-on-unhealthy` and try to find a way of removing inheritance from the Dockerfile syntax (not too easy - it uses the same mechanism for command-line overrides as for image overrides so we still need to support inheritance in the implementation but disable it for images).

aluzzardi commented:
/cc @dongluochen

aluzzardi commented:
> We don't want to handle failure cases within Docker right now, only collect and expose the information in the API

I think this would make the feature substantially useless.

If I have to write software to periodically poke the engine API in order to gather the health check status and take action, I might as well just run the health checking command myself at this point.

crosbymichael commented:
@aluzzardi I'm not sure and the point was that this initial PR lays the groundwork for building the logic for acting upon the results in a second PR.  

I feel like collecting and query is half of the problem and it would be good to encode that information in images letting the authors decide that it means for their Image to be healthy.  

The arguments were more towards lets get the collection and query done correct, where is can be used as a standalone feature as not everyone will want the single exit on failure and will have to query periodically anyways.

tianon commented:
I agree -- collecting the information in the daemon is really great to keep separate from acting on changes in that state (since those are more controversial and have more varying use cases).  On the other side, I think tracking and reporting state changes as container events makes good sense (such that applications consuming this early implementation can do so without polling, which I do agree would make this somewhat of a moot feature otherwise).

I'd also add that if we've taken the default "action" out of the health checks, then as long as we have a way for dependent images to clear any "heath check", inheritance actually _does_ make sense.  In my mind, having the image decide what actions should be performed on the running container in response to changes in heath status is in the same category as letting the container decide that it requires (and force-enables) `--privileged` or `--tty` -- those are runtime/image operator concerns for which the image builder ought only to be making recommendations.

crosbymichael commented:
Ya, adding these different statuses to the event stream makes alot of since as there are already meany systems listening for load balancing and other systems.

thaJeztah commented:
I like the "health" event idea

mgoelzer commented:
I agree with @aluzzardi .  Either way, is it necessary to remove `--exit-on-unhealthy` completely?  All of this stuff is overridable on the command line, no?

@tianon I see inheritance as an orthogonal issue to this.  I think we should drop it because it's confusing and counterintuitive.

stevvooe commented:
> Either way, is it necessary to remove --exit-on-unhealthy completely? All of this stuff is overridable on the command line, no?

This issue is that this has nothing to do with a health check. This is a policy _based_ on the health check. I think it is fine that we have it -- it just needs to be integrated with the restart policy.

dongluochen commented:
> This issue is that this has nothing to do with a health check. This is a policy based on the health check. I think it is fine that we have it -- it just needs to be integrated with the restart policy.

I think it a good idea to put action inside restart policy to make health check as non-intrusive as possible. This first iteration may provide data for us to decide defaults and actions.

Commit:
Print a note when overriding a HEALTHCHECK from the base image

Requested by Mickael Laventure.

Signed-off-by: Thomas Leonard <thomas.leonard@docker.com>

Commit:
Remove --exit-on-unhealthy

It's not yet clear what the design for this should be, so remove it for
now.

Signed-off-by: Thomas Leonard <thomas.leonard@docker.com>

Commit:
Skip health check integration test on Windows

Signed-off-by: Thomas Leonard <thomas.leonard@docker.com>

Commit:
Implement SignalProcess on Windows

Untested.

Signed-off-by: Thomas Leonard <thomas.leonard@docker.com>

Commit:
Allow probe to request staying in the "starting" state

This is useful for e.g. a database that is performing a long recovery
process. It allows the container to exceed the grace period without
having to become "healthy" or "unhealthy".

Also, added unit-test for state transitions.

Signed-off-by: Thomas Leonard <thomas.leonard@docker.com>

Commit:
Remove --grace healthcheck option

Scripts can now handle this themselves now by returning with exit code 2.
Avoids hard-coding the timeout in the image, which is impossible to do
correctly as it is sensitive to the speed of the target platform (and
therefore not portable).

Signed-off-by: Thomas Leonard <thomas.leonard@docker.com>

Commit:
Represent durations as integer nanoseconds

This simplifies the code. Also, now that grace isn't an option zero is
no longer a sensible value, so we can use 0 to mean unspecified and
avoid using a pointer type.

Signed-off-by: Thomas Leonard <thomas.leonard@docker.com>

Commit:
Report the last few healthcheck results

Signed-off-by: Thomas Leonard <thomas.leonard@docker.com>

talex5 commented:
I've made some updates based on the feedback here and from talking to a few people:

- The `--grace` option is gone. Instead, the probe command should return an exit status of 2 to indicate that the container is still starting (e.g. if it sees the server hasn't written its PID file yet). This avoids baking a fixed timeout into the image, which isn't portable (different machines run at different speeds) and doesn't cope with variable startup times (e.g. a database performing recovery after a crash).

- `--exit-on-unhealthy` is gone for now. It's not clear how this should work and we can add this in a later PR.

- The output of `docker inspect` shows the last few results, not just the last one.

- If you override a healthcheck from a base image, an informational note is printed.

Commit:
Change some types

Requested by Stephen Day.

Signed-off-by: Thomas Leonard <thomas.leonard@docker.com>

aluzzardi commented:
> The --grace option is gone. Instead, the probe command should return an exit status of 2 to indicate that the container is still starting (e.g. if it sees the server hasn't written its PID file yet). This avoids baking a fixed timeout into the image, which isn't portable (different machines run at different speeds) and doesn't cope with variable startup times (e.g. a database performing recovery after a crash).

I think that would force probes to be shell wrappers (e.g. users can't just run `curl` or `redis-cli` etc).

Also, as a probe I have no idea if I'm within the grace period or not.

For instance, in the case of a redis probe, I would need to create a wrapper around `redis-cli`. If that command fails, I have no idea if I should return 1 or 2 - is redis non responding because it's booting up or because it crashed?

talex5 commented:
@aluzzardi I think you could just check whether the PID file exists. If not found, then redis is still starting. If that doesn't work, perhaps the `supervised` option could be used to write a file that the health probe checks. If nothing else works, I guess you could look at the start time of PID 1 to see how long it has been running and make a decision based on that.

Note: I made the commit that removes grace separate to the one that adds suppport for the probe asking for more time (by returning `-2`). So perhaps we could merge this now and discuss adding the grace feature back as a separate PR if no other solution can be found?

talex5 commented:
(note: the vendor test is failing because https://github.com/docker/engine-api/pull/236 isn't merged yet)

stevvooe commented on vendor/src/github.com/docker/engine-api/types/container/config.go:
```diff
@@ -1,10 +1,31 @@
+	// {"NONE"} : disable healthcheck
+	// {"CMD", args...} : exec arguments directly
+	// {"CMD-SHELL", args...} : run command with system's default shell
+	Test strslice.StrSlice `json:",omitempty"`
```
Glad I just caught this: `strslice.StrSlice` is for backwards compatibility. Just use `[]string`. There is not need to support string commands for new stuff.

stevvooe commented on vendor/src/github.com/docker/engine-api/types/types.go:
```diff
@@ -274,6 +274,19 @@ type ExecStartCheck struct {
+type HealthcheckResult struct {
+	Start, End time.Time
+	ExitCode   int    // 0=healthy, 1=unhealthy, 2=starting
+	Output     string // Output from last check
```
`[]byte`?

stevvooe replied:
Actually, this has to be `[]byte`, so it gets armored against bad or binary output.

aaronlehmann replied:
That would result in go base64 encoding this value, which means the value would not be human readable in the JSON. IMHO it would be better to use a string, and accept that binary output will look like garbage.

stevvooe replied:
Will Go's json package appropriate escape the content? Will all json libraries be able to recover that escaping correctly?

aaronlehmann replied:
I can't make any guarantees about all json libraries, but the json grammar is very unambiguous that all unicode characters except `"`, `\`, and control characters are valid in strings. Go will escape those exceptions.

stevvooe replied:
In Python 2.7:

```python
>>> s = ''.join(chr(i) for i in range(0,256))
>>> s
'\x00\x01\x02\x03\x04\x05\x06\x07\x08\t\n\x0b\x0c\r\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f !"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~\x7f\x80\x81\x82\x83\x84\x85\x86\x87\x88\x89\x8a\x8b\x8c\x8d\x8e\x8f\x90\x91\x92\x93\x94\x95\x96\x97\x98\x99\x9a\x9b\x9c\x9d\x9e\x9f\xa0\xa1\xa2\xa3\xa4\xa5\xa6\xa7\xa8\xa9\xaa\xab\xac\xad\xae\xaf\xb0\xb1\xb2\xb3\xb4\xb5\xb6\xb7\xb8\xb9\xba\xbb\xbc\xbd\xbe\xbf\xc0\xc1\xc2\xc3\xc4\xc5\xc6\xc7\xc8\xc9\xca\xcb\xcc\xcd\xce\xcf\xd0\xd1\xd2\xd3\xd4\xd5\xd6\xd7\xd8\xd9\xda\xdb\xdc\xdd\xde\xdf\xe0\xe1\xe2\xe3\xe4\xe5\xe6\xe7\xe8\xe9\xea\xeb\xec\xed\xee\xef\xf0\xf1\xf2\xf3\xf4\xf5\xf6\xf7\xf8\xf9\xfa\xfb\xfc\xfd\xfe\xff'
>>> json.dumps(s)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/__init__.py", line 243, in dumps
    return _default_encoder.encode(obj)
  File "/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/encoder.py", line 201, in encode
    return encode_basestring_ascii(o)
UnicodeDecodeError: 'utf8' codec can't decode byte 0x80 in position 128: invalid start byte
```

In this example, we attempt to encode a full string with all byte values into json. This fails, since python tries to coerce the string to unicode before encoding the json. This fails because the binary data is not valid unicode.

aaronlehmann replied:
Yes, but my understanding is that Go's json encoder will escape any values that are not valid unicode.

The problem arises when you try to decode a JSON string that's not properly escaped. But Go should apply the correct escaping rules.

stevvooe replied:
This doesn't even round trip in Go:

```
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"reflect"
)

func main() {
	var b []byte
	for i := 0; i < 256; i++ {
		b = append(b, byte(i))
	}

	type V struct {
		S string
		B []byte
	}

	out := V{string(b), b}
	p, err := json.Marshal(out)
	if err != nil {
		log.Fatalln(err)
	}

	os.Stdout.Write(p)
	fmt.Println()

	var v V
	if err := json.Unmarshal(p, &v); err != nil {
		log.Fatalln(err)
	}

	var fail bool
	if v.S != out.S {
		fail = true
		log.Println("strings do not match")
	}

	if !reflect.DeepEqual(v.B, out.B) {
		fail = true
		log.Println("bytes do not match")
	}

	if !reflect.DeepEqual(v, out) {
		fail = true
		log.Println("input does not match output")
	}

	if fail {
		log.Fatalln("input and output differ")
	}

}

```

This outputs:

```go
go run foo.go
{"S":"\u0000\u0001\u0002\u0003\u0004\u0005\u0006\u0007\u0008\t\n\u000b\u000c\r\u000e\u000f\u0010\u0011\u0012\u0013\u0014\u0015\u0016\u0017\u0018\u0019\u001a\u001b\u001c\u001d\u001e\u001f !\"#$%\u0026'()*+,-./0123456789:;\u003c=\u003e?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd","B":"AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8gISIjJCUmJygpKissLS4vMDEyMzQ1Njc4OTo7PD0+P0BBQkNERUZHSElKS0xNTk9QUVJTVFVWV1hZWltcXV5fYGFiY2RlZmdoaWprbG1ub3BxcnN0dXZ3eHl6e3x9fn+AgYKDhIWGh4iJiouMjY6PkJGSk5SVlpeYmZqbnJ2en6ChoqOkpaanqKmqq6ytrq+wsbKztLW2t7i5uru8vb6/wMHCw8TFxsfIycrLzM3Oz9DR0tPU1dbX2Nna29zd3t/g4eLj5OXm5+jp6uvs7e7v8PHy8/T19vf4+fr7/P3+/w=="}
2016/05/26 13:37:52 strings do not match
2016/05/26 13:37:52 input does not match output
2016/05/26 13:37:52 input and output differ
```

This _must_ be a `[]byte`.

talex5 replied:
Why would it round-trip? If the script generates invalid UTF-8 then Go will replace the bad data with unicode replacement characters. The probe output is intended to be displayed as text, not used as some kind of reliable binary transport.

stevvooe replied:
> Why would it round-trip?

Usually, when we send it over engine API.

> Go will replace the bad data with unicode replacement characters

What constitutes "bad data"? This is the output of a command. What if the output is latin-1 or gb2312? What if it is image data or a binary data?

talex5 replied:
It's limited to 255 bytes, so images would be tricky. I've updated the docs to specify UTF-8: https://github.com/talex5/docker/blob/healthcheck/docs/reference/builder.md#healthcheck

> To help debug failing probes, any output text (UTF-8 encoded) that the command writes on stdout or stderr will be stored in the health status and can be queried with docker inspect. Such output should be kept short (only the first 255 bytes are stored currently).

stevvooe replied:
I really don't understand what the purpose of limiting this is. You're overcomplicating the validation and support when a simple `[]byte` will take care of most of this. Sure output can be truncated, but it is senseless for us to care about the encoding.

justincormack replied:
If you truncate it you do have to care about the encoding, or you might truncate in the middle of a UTF8 sequence and output invalid bytes later, so stupid as it is I think it might be best to use a string, though I would probably recommend ascii output.

stevvooe commented on vendor/src/github.com/docker/engine-api/types/types.go:
```diff
@@ -274,6 +274,19 @@ type ExecStartCheck struct {
+// Health stores information about the container's healthcheck results.
+type Health struct {
+	Status        string               // States are: "starting", "healthy", "unhealthy"
+	FailingStreak uint64               // Number of consecutive failures
```
`int`.

stevvooe commented on vendor/src/github.com/docker/engine-api/types/container/config.go:
```diff
@@ -1,10 +1,31 @@
+
+	// Number of consecutive failures needed to consider a container as unhealthy.
+	// Zero means inherit.
+	Retries uint `json:",omitempty"`
```
`int`.

Commit:
Move health states to engine-api

Signed-off-by: Thomas Leonard <thomas.leonard@docker.com>

Commit:
Clarify that the probe's debug output is UTF-8 text

Signed-off-by: Thomas Leonard <thomas.leonard@docker.com>

crosbymichael commented on api/server/router/container/backend.go:
```diff
@@ -17,7 +17,7 @@ type execBackend interface {
 	ContainerExecInspect(id string) (*backend.ExecInspect, error)
 	ContainerExecResize(name string, height, width int) error
-	ContainerExecStart(name string, stdin io.ReadCloser, stdout io.Writer, stderr io.Writer) error
+	ContainerExecStart(ctx context.Context, name string, stdin io.ReadCloser, stdout io.Writer, stderr io.Writer) error
```
Is this supposed to be here?

talex5 replied:
Yes. If a healhcheck probe hangs, we need a way to kill it.

dannc commented:
Would this be accessible to run a health check separately via compose?

If I have some arbitrary image that my container needs to have "running" (or completed and exited) before I start up, I often won't have any idea if it's implemented a proper health check if I'm getting it from a third-party source (such as DockerHub).  This would lead me to have to:

* extend the image I rely on to add in a (possibly redundant) health check, or
* modify the startup logic of my image to check whether the image I depend on has "completed" for my purposes, or
* in the case where my dependent image is itself from a third-party source, I would have to extend that and check the status of the image I depend on.

Having the ability to put a health check in my compose file (which would function like docker run startup arguments) would make it much more explicit and I'd be less likely to need to extend images that I'm using for this purpose.

crosbymichael commented:
Design LGTM

aluzzardi commented:
> @aluzzardi I think you could just check whether the PID file exists. If not found, then redis is still starting. If that doesn't work, perhaps the supervised option could be used to write a file that the health probe checks. If nothing else works, I guess you could look at the start time of PID 1 to see how long it has been running and make a decision based on that.

Yes, but that's exactly what I was implying: now health checks have to be shell scripts since a simple CLI command is not enough.

We already have `entrypoint.sh` in virtually every base image and we're going to add `healthcheck.sh`

stevvooe commented on daemon/health.go:
```diff
@@ -0,0 +1,314 @@
+}
+
+// Buffer up to maxOutputLen bytes. Further data is discarded.
+type limitedBuffer struct {
```
Surely, this already exists in the stdlib. Please don't reimplement.

talex5 replied:
There is an `io.limitedReader`, but no `io.limitedWriter`.

stevvooe replied:
Unfortunate. `ContainerExecStart` is backwards...

stevvooe commented on vendor/src/github.com/docker/engine-api/types/types.go:
```diff
@@ -274,6 +274,27 @@ type ExecStartCheck struct {
+type HealthcheckResult struct {
+	Start, End time.Time
+	ExitCode   int    // 0=healthy, 1=unhealthy, 2=starting, -1=error running probe
+	Output     string // Output from last check
```
I'm sorry to bring this up again: this is wrong. This needs to be a `[]byte` and should allow a more than 255 bytes of output. There is simply no good reason for these kinds of limitations.

There are lots of things that might go into a health check result output that you haven't thought of. This is going to be especially true when compound script checks may want to report on a number of parameters.

talex5 replied:
If so, we can raise the limits later when we add that. Allowing containers to store unbounded amounts of data in Docker's JSON status files (and return it in inspect replies) would make DoS attacks very easy. Allowing binary data prevents tools from displaying the output to the user in a readable form. It would be better to add a separate field later if we decide on some structured payload format (and not mix it with error output).

stevvooe replied:
Look, without a healthcheck protocol, assuming anything about the output of a healthcheck is silly. What if there are several aspects to health check that need to be displayed? What if someone wants to use a binary protocol for the healthcheck? What about debugging?

The salient portion of the healthcheck my not even be in the first 255 bytes.

I'm sorry, but this is a ridiculous limitation.

talex5 replied:
This PR is defining the healthcheck protocol, and so we can define it to be whatever we want.

stevvooe commented on docs/reference/builder.md:
```diff
@@ -1300,6 +1300,73 @@ The `STOPSIGNAL` instruction sets the system call signal that will be sent to th
+    HEALTHCHECK --interval=5m --timeout=3s \
+      CMD curl -f http://localhost/ || exit 1
+
+To help debug failing probes, any output text (UTF-8 encoded) that the command writes
```
This is an extremely silly limitation.

talex5 replied:
Without specifying the encoding, tools that consume the healthcheck data (e.g. UCP or the docker cli) have no way to display it to the user.

stevvooe replied:
We do this every day with shell commands without these limitations and it works just fine. Sanitization is left for the destination, not the source. Putting this into UTF-8 is not some sort of magic armor for consumers of this output. This conversion can be done at the CLI or UCP, where it can be secure and tailored for the target.

An interesting case here is CJK users. Using GB2312, you can get about 127 characters in 255 bytes. Forcing UTF-8 limits them to 85 characters. By forcing an encoding and size, you're limiting information density achievable in other languages.

This also prevent existing tools from working that aren't careful about ensuring UTF-8 output. Do you even have guidelines on how to ensure correct UTF-8 via standard out? In python? In ruby? In C?

Again, this is a ridiculous and arbitrary limitation.

stevvooe commented on vendor/src/github.com/docker/engine-api/types/types.go:
```diff
@@ -274,6 +274,27 @@ type ExecStartCheck struct {
+
+// Health states
+const (
+	Starting  = "starting"  // Container is not yet ready
```
If a container transitions to `healthy`, then fails one health check, what is the next state to report? If a health check is run several times, how does it know its last state?

talex5 replied:
Assuming `Retries=1`, the state after one failure would be `unhealthy`.

Usually, a healthcheck does not need to know its previous state in order to report its current state.

A possible exception might be a very simple probe that detects failure, but wants to report "starting" if it's currently still in the starting state (because it can't distinguish between starting and unhealthy). However, it can get this behaviour by returning exit code 2 in all cases, since "starting" is treated as unhealthy if the container wasn't already starting.

tianon replied:
It seems to me that for the first rendition of this feature, it'd be simpler and offer more flexibility for the future if we only have two states, `healthy` or `unhealthy`, and that state should be based directly on the zero or non-zero exit code of the most recent health check, and leave more complex states (grace period during startup, grace period for intermittent failures, etc) for further discussion once we have a good feeling for exactly what these transitions look like in the real world, and how this feature is going to be used longer-term.

talex5 commented:
@dannc Yes, you can specify a healthcheck for an existing image with `docker run`, and compose could use the same API I imagine.

stevvooe commented:
👎 until we remove the arbitrary output limitations.

aluzzardi commented:
@stevvooe Unfortunately, I'm not sure that would work.

The output definition I see here is for the engine-api, which is JSON encoded. I don't think you can have a `[]byte` there.

crosbymichael commented:
You can have a `[]byte` but it gets base64 encoded by the marshaler.  

I think you would need some type of limit on the output that we save or else it could cause problems.  @stevvooe what do you suggest?  a larger limit or a better defined protocol about what output we will collect?

stevvooe commented:
> You can have a []byte but it gets base64 encoded by the marshaler.
> I don't think you can have a []byte there.

Encoding to base64 is by design, so output is reproduced faithfully. Go [round trips this transparently](https://github.com/docker/docker/pull/22719#discussion_r64816770) where string loses data when not encoded as UTF-8. There is more about this [here](https://groups.google.com/forum/#!topic/golang-nuts/kXMWIfSTix0).

The main issue here is that it will be extremely hard to track down the source of the incorrect bytes when debugging a problem.

I would suggest an output limit of 4096. It is still arbitrary but not limiting.

This is probably okay to enforce for `docker inspect` output, but let's cast to utf-8 and truncate there. I just don't see how to troubleshoot health checks when the output is bigger or there is an encoding problem.

thaJeztah commented:
> This is probably okay to enforce for docker inspect output, but let's cast to utf-8 and truncate there

I can see that as a possibility; my concern was about `docker inspect` being (without additional tooling) useless if the output is `base64` encoded. We could still output both ("raw" and "readable")

aluzzardi commented:
> You can have a []byte but it gets base64 encoded by the marshaler.

That's true in Go but it's very language specific.

What I meant by "you can't have a []byte there" is, at the end, that field is going to be a string no matter what. Whether we base64 encode or not is another question.

How do we currently encode output for other commands (e.g. logs)?

stevvooe commented:
What if we did `Raw` with `[]byte` and `Output` with string?

thaJeztah commented:
> What if we did Raw with `[]byte` and `Output` with string?

SGTM, would give both options; would `Raw` be limited in size?

thaJeztah commented:
ping @aluzzardi @crosbymichael WDYT on @stevvooe's proposal?

Commit:
Remove the arbitrary output limitation of 255 bytes

Use 4096 instead.

Signed-off-by: Thomas Leonard <thomas.leonard@docker.com>

talex5 commented:
I'm not sure what you mean about having both. Would we include the same data twice in each message, or someone know where to send which data?

Another possibility: if a probe wants to generate binary data then it base64 encodes it itself. That way, probes that only want text don't need to do anything and the messages are human-readable. Probes that return binary data would need something at the client end to interpret it anyway, so decoding it wouldn't be much extra work. The data on the wire would be the same as if we'd used `[]byte` in that case.

stevvooe commented:
@talex5 Yes, have both. One that is "human-readable", and hopefully properly sanitized, and one that is the raw, unprocessed output.

This will meet both the use cases of sending binary data and debugging problems with health check output during development. It is unreasonable to make users go through so much work just to get unprocessed command output.

Let's make sure this feature isn't limited to _our_ imaginations but, instead, enables others'.

cpuguy83 commented:
I do not see the purpose of supporting more than just simple output from a healthcheck, and certainly not storing both binary and human-readable....

I think we should support only something very simple for a human to use as debug info, or potentially remove support for capturing output from the healthcheck from this PR and implement separately.

cpuguy83 commented on docs/reference/builder.md:
```diff
@@ -1300,6 +1300,73 @@ The `STOPSIGNAL` instruction sets the system call signal that will be sent to th
+
+- 0: success - the container is healthy and ready for use
+- 1: unhealthy - the container is not working correctly
+- 2: starting - the container is not ready for use yet, but is working correctly
```
Isn't this state really just any failure before the first `success`?
I would recommend removing it.

talex5 replied:
Without this, whatever is monitoring the status will need some timeout to decide that the container has failed to start. @justincormack (I think) pointed out that a simple timeout is not enough to cover e.g. a database that is busy performing recovery actions. So the idea of this state is to allow the container to tell the monitor that it needs more time, and avoid the monitor having to guess what the grace period should be.

dannc replied:
A nit about the labels/description attached to the return codes.  Some containers are "run once and die" in that they produce some artifact and exit.  Take for example a container that generates a TLS certificate and writes it to a volume, which other containers subsequently mount and use the certificate.

I would propose something like:
0: succeeded - the container is healthy or has exited successfully
1: failed - the container is unhealthy or has exited unsuccessfully
2: starting (in-progress?) - the container is not ready for use yet or has not completed its task

cpuguy83 commented on docs/reference/builder.md:
```diff
@@ -1300,6 +1300,73 @@ The `STOPSIGNAL` instruction sets the system call signal that will be sent to th
+
+The options that can appear before `CMD` are:
+
+* `--interval=DURATION` (default: `30s`)
```
Did we really decide to keep these runtime related flags?

talex5 replied:
We agreed to get rid of grace, but I'm happy to remove the others too.

crosbymichael commented:
For the output, in this PR we should keep it as is.  The size has already been increased to 4096 which is much better than what it was before.  Most commands that do health checks should be encouraged to write a simple reason why they are reporting as they did.  "I returned unhealthy BECAUSE i could not connect to the database."  We don't want to encourage checks to do a yolo operation and puke up some stacktrace to decipher.  

```c
println("could not connect to database");
return 1;
```

In the future after this has some user and they are wanting some type of more structured response than a limited string it would be effortless to add an unbounded []byte array to the output for this.  It is something that is safe to defer until we gain feedback.

For the three states, simple health checks don't have to use the starting state if they don't want to.  To the normal developer they can just return 1 or 0.  If the app has a long startup or some type of complexity on boot then it can take advantage of this state and we don't have to infer any type of guesses that we would otherwise be doing.  Lets to write software that guesses things, let the things that know the state tell us.  You are not forced to use this state if you don't want to.

After looking at the code again, I think all these things are already implemented.

LGTM

thaJeztah commented:
ok, enough bike-shedding. Let's go for it LGTM

dongluochen commented on builder/dockerfile/parser/testfiles/health/Dockerfile:
```diff
@@ -0,0 +1,10 @@
+ADD check.sh main.sh /app/
+CMD /app/main.sh
+HEALTHCHECK
+HEALTHCHECK --interval=5s --timeout=3s --retries=1 \
```
nit: I think a good practice is multiple failures like 3 to tolerate random failures. Giving an example of `--retries=1` might not be appropriate.

dongluochen replied:
I suggest we set default retries to 3.

dongluochen commented:
PR looks good to me.

thaJeztah commented:
@dongluochen I carried the PR, rebased/squashed and with vendor update in https://github.com/docker/docker/issues/23218. Let me know if you need those addressed in the PR, otherwise I'll do a follow-up

dongluochen commented:
Thanks @thaJeztah. Follow-up is good.

mageddo commented:
Guys what about the restart policy? I think that is very important, the reason was explained here. So will this feature be implemented?

stevvooe commented:
@mageddo This is a long closed pull request, carried elsewhere. If you have questions or a feature request, please file a new issue.

thaJeztah commented:
Locking the conversation in this issue for the reason @stevvoe mentioned above; comments on closed issues and PRs easily go unnoticed - I'm locking the conversation to prevent that from happening
